dynamic_programming/unique_paths.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * @file
+ * @brief Implementation of Unique Paths problem using Dynamic Programming.
+ * @details
+ * A robot is located at the top-left corner of an m × n grid.
+ * The robot can move either down or right at any point in time.
+ * This program computes the total number of unique paths to reach
+ * the bottom-right corner.
+ *
+ * Approaches:
+ * - **Memoization (Top-Down)**: Recursively explores solutions while
+ *   storing intermediate results in a cache (`dp`) to avoid redundant
+ *   computation. Typically more intuitive and easy to write, but
+ *   relies on recursion and has associated call-stack overhead.
+ *
+ * - **Tabulation (Bottom-Up)**: Iteratively builds the solution using
+ *   a DP table, starting from the base cases and filling up the table.
+ *   Offers consistent performance without recursion overhead and
+ *   is generally more space-efficient when optimized.
+ *
+ * @see https://leetcode.com/problems/unique-paths/
+ */
+
+#include <iostream>
+#include <vector>
+#include <cassert>
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+
+/**
+ * @class UniquePathsSolver
+ * @brief Solves the Unique Paths problem using both memoization and tabulation.
+ */
+class UniquePathsSolver {
+   private:
+    std::vector<std::vector<int>> dp;  ///< Memoization table
```

Comment:
this variable name is too short, its nondescript/confusing. maybe rename it to memoization_table or something similar